Exercise 3, first version

muzaluisa opened the pull request:
Hey all,

Will be thankful for your feedback on the second exercise :)
I used the example, given in the assignment. Used queues and width-first-search to sebsequently add new routes and prune non-perspective ones (not from prefixes). 

Last time I was not able to review other students homework., but now I am going to do that :)

muzaluisa committed:
Exercise 3, first version

muzaluisa committed:
Create codeU3.py

Improved docstrings

luxiya01 commented on codeU3.py:
```diff
@@ -0,0 +1,222 @@
+    '''
+
+    return set([word[0:n] for n in range(1, len(word)+1)])
+
```
I really like your solution about getting prefixes! However, perhaps it can be a class function within the class Dictionary instead?

muzaluisa replied:
Good point, done)

luxiya01 commented on codeU3.py:
```diff
@@ -0,0 +1,222 @@
+        if word.lower() in self.words:
+            return True
+        return False
+
```
I think you can just return "word.lower() in self.words"? 
It's probably nice with the error message saying that the input needs to be a string, but since python doesn't really care about type, maybe it's not really necessary?

muzaluisa replied:
I can add the test case indeed, when the input is a number, good point)

luxiya01 commented on codeU3.py:
```diff
@@ -0,0 +1,222 @@
+    def test_nonstring_input(self):
+        self.assertRaises(self.d.isWord(['cat', 'cats']))
+        self.assertRaises(self.d.isPrefix(['cat', 'cats']))
+
```
Maybe it's easier to navigate the codes if you split one file per class? And maybe not mix test with other classes. (Or it could also just be my Java habit TAT. )

blmussati replied:
I was thinking the same, it makes it easier to read to have class files, test files and function files separated :) but the code looks good, so this is just a matter of simplicity!

muzaluisa replied:
fixed that)

luxiya01 commented on codeU3.py:
```diff
@@ -0,0 +1,222 @@
+        prefix = ''.join(chars)
+        if self.dict.isPrefix(prefix):
+            return True
+        return False
```
Here you can probably return "self.dict.isPrefix(prefix)" directly?

muzaluisa replied:
True :) Fixed!

luxiya01 commented on codeU3.py:
```diff
@@ -0,0 +1,222 @@
+                    words_found.add(str(word))
+                if len(route) < self.dict.max_word_len:
+                    routes.extendleft(self.route_extend(route))
+        return words_found
```
I really like the use of deque!

siimphh replied:
Yes, works nicely here :)

I wonder though if there is a  reason you are avoiding recursion? It's a fine choice - possibly making it easier to read - just interested about the background.

muzaluisa replied:
I agree that the recursion approach is quite natural here, just generally use recursion carefully, due to possible time overheads :)

luxiya01 commented on codeU3.py:
```diff
@@ -0,0 +1,222 @@
+            if self.is_route_prefix(route):
+                word = self.route_word(route)
+                if word:
+                    words_found.add(str(word))
```
You seem to be computing "word" for quite many times here. Don't know if it matters at all, but perhaps one could consider storing the current string directly and use isPrefix and isWord directly to check whether the string given by the route so far is valid?

siimphh replied:
Indeed, it might be a bit cleaner to do:

```
word = self.route_to_word(route)
if self.dict.is_prefix(word):
  if self.dict.is_word(word):
    ...
```

muzaluisa replied:
I agree, made changes accordingly!)

siimphh commented on codeU3.py:
```diff
@@ -0,0 +1,222 @@
+
+    def __init__(self, word_list):
+        if not word_list:
+            assert ValueError, 'Please provide non-empty list of words'
```
Exceptions (such as `ValueError`) in Python are raised as:

```
raise ValueError('Please provide ...')
```

Assertions are more concise in this case:

```
assert word_list, 'Please provide ...'
```

Which would already include and thus would not require the `if not word_list:` above (the first term of an assertion is the check).

muzaluisa replied:
That is indeed better :) Fixed!

siimphh commented on codeU3.py:
```diff
@@ -0,0 +1,222 @@
+        if not word_list:
+            assert ValueError, 'Please provide non-empty list of words'
+        word_list = [w.lower() for w in word_list if w]
+        self.max_word_len = reduce(max, map(len, word_list))
```
Can just be `max(map(len, word_list))`.

siimphh commented on codeU3.py:
```diff
@@ -0,0 +1,222 @@
+
+        if not isinstance(word, str):
+            assert TypeError, 'Input should be a string'
+            return False
```
This `return` statement is never reached - both `assert` and `raise` break the flow already.

muzaluisa replied:
Agree, fixed, somehow I did not get used to use assert, rather exceptions, now I understand the difference.

siimphh commented on codeU3.py:
```diff
@@ -0,0 +1,222 @@
+
+class CharGrid(object):
+
+    '''Given (n by m) grid of letters and a dictionary
```
Thanks for writing a class-level docstring as well as the function docstrings - helps a lot with reading and understanding any code base!

siimphh commented on codeU3.py:
```diff
@@ -0,0 +1,222 @@
+
+        if not route:
+            return ""
+        chars = [self.grid[x, y] for (x, y) in route]
```
Might be worth making it a `route_to_word(route)` function as it occurs the second time now.

muzaluisa replied:
yes, this function is useful, made that :)

siimphh commented on codeU3.py:
```diff
@@ -0,0 +1,222 @@
+        '''Given a list of tuple coordinates (x,y) encoding route
+        Returns:
+            empty string, if corresponding string is not in the dictionary
+            word, otherwise
```
Partial docstring!

siimphh commented on codeU3.py:
```diff
@@ -0,0 +1,222 @@
+            word, otherwise
+        '''
+
+        if not route:
```
If the route is an empty list then the code below would just work so this special case is not required?

muzaluisa replied:
Not required, fixed that )

siimphh commented on codeU3.py:
```diff
@@ -0,0 +1,222 @@
+        cells_set = set(route)
+        route_extended = []
+
+        if last_x + 1 < self.m and (last_x + 1, last_y) not in cells_set: #ensure we don't visit the same cell twice
```
It could be the right choice but most people would also consider some kind of loop for this. Do you have opinions why it's best to fully write all this conditions out?

muzaluisa replied:
I have made a new version with a loop that is in fact indeed shorter :) Fully writing the conditions in fact is a waste of space, but was easier for me in the first place to code (more straight forward)  
```
for i in range(-1, 2):
            for j in range(-1, 2):
                if i == j and i == 0:
                    continue
                if 0 <= last_x + i < self.m and 0 <= last_y + j < self.n\
                and (last_x + i, last_y + j) not in cells_set:
                    route_extended.append(route + [(last_x + i, last_y + j)])
```

siimphh replied:
Yep, I also find that sometimes it is helpful just to write some code that gets the job done. Working slightly sub-optimal code is better than perfectly optimal code that is not written yet :)

siimphh commented on codeU3.py:
```diff
@@ -0,0 +1,222 @@
+        all_cells = [[(i, j)] for i in range(self.m) for j in range(self.n)]
+        routes = deque(all_cells)
+        words_found = set()
+        while routes:
```
You've chosen to enumerate all potential words and compare them with a dictionary. The other option would be to take words from a dictionary and look for them in the grid.

Are the advantages to either option?

muzaluisa replied:
The choice of the approach depends on the size of the grid and the amount of words in the vocabulary indeed. So my solution is optimized for large vocabulary, but less for larger grids. Knowing the largest length of the word M helps to stop the traversal earlier in case larger grids though. If one word is super-long, that will decrease the performance. 

Alternative approach: On the other hand if the vocabulary is small, for each word we can explore several number routes of max len of this word M_i, that is BFS/DFS search

siimphh replied:
I would tend to be more worried about the number of words expressed by the grid for any grid of non-trivial size. I imagine 10x10 grid would contain effectively infinite words (or maybe not?). So while a word list can be long, even counting the number of words in the grid is a challenge!

siimphh reviewed (commented):
Nice solution, was very easy to read and follow, thanks for your effort in structuring, formatting and documenting the code!

muzaluisa committed:
Refined by the first review round

Comments are provided by @windo, @luxiya01  and @blmussati , thank you for your reviews!:)
Could you please look again modified version?

muzaluisa reviewed (commented):
Could you please check an updated version with all your feedback taken into account?
